File: pset2/vigenere.c
/*
 vigenere.c
 
 Brandan McDevitt
 Harvard Computer Science 50
 Week 2 Problem Set
 
 Implement a program that encrypts messages using Vigenère’s cipher.
 */

#include <stdio.h>
#include <cs50.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

//declaring variables
string key;
string message;
int kLength;

int main(int argc, string argv[])
{
    //if the input count is not equal to 2 then inform the user and end
    if (argc != 2)
    {
        printf("Incorrect usage. Try again.\n");
        printf("./vigenere key\n");
        return 1;
    }
    //else if the count is equal to 2 but there are non alphabetical letters, inform the user and end
    else
    {
        for (int i = 0; i < strlen(argv[1]); i++)
        {
            if (isalpha(argv[1][i]) == false)
            {
                printf("Incorrect usage. Try again.\n");
                printf("Key must be alphabetical.\n");
                return 1;
            }
        }
    }

    //assigning the key and length of the key to variables
    key = argv[1];
    kLength = strlen(key);

    //storing the message from the user
    message = get_string("Enter your message to be encrypted: ");

    printf("ciphertext: ");

    //a loop for going through each character and replacing it with the encrypted letter
    for (int i = 0, j = 0; i < strlen(message); i++)
    {
        int letterKey = tolower(key[j % kLength]) - 'a';

        if (isupper(message[i]))
        {
            printf("%c", 'A' + (message[i] - 'A' + letterKey) % 26);

            j++;
        }
        else if (islower(message[i]))
        {
            printf("%c", 'a' + (message[i] - 'a' + letterKey) % 26);
            j++;
        }
        else
        {
            printf("%c", message[i]);
        }
    }

    printf("\n");

    return 0;

}

File: pset6/vigenere/vigenere.py
# vigenere.py

# Brandan McDevitt
# Harvard Computer Science 50
# Problem Set 6

# Port vigenere.c to python.

from cs50 import get_string
from sys import argv

# check input to make sure that the user entered a valid key
if len(argv) != 2:
    print("Incorrect usage. Try again. \ni.e. python vigenere.py <key>")
    exit(1)
else:
    for letter in range(len(argv[1])):
        if argv[1][letter].isalpha() == False:
            print("Incorrect usage. Try again. \n<key> must be alphabetical")
            exit(1)

key = []

# loop through the characters in the key and append them to the key[] list
for character in argv[1]:
    if character.isupper():
        character = (ord(character)) - (ord('A'))
        key.append(character)
    elif character.islower():
        character = (ord(character)) - (ord('a'))
        key.append(character)

# getting the length of the key
keyLength = len(key)

# prompt user for the message
message = get_string("Enter your message to be encrypted: ")

print("ciphertext: ", end="")

character = 0
i = 0

# loop through the characters in the message and replace the value dependent on the key
for character in message:

    # validate whether the character is alphabetical
    if character.isalpha():
        # check if the character is uppercase
        if character.isupper():
            character = ((ord(character) - ord('A') + (key[i])) % 26) + ord('A')
            print(chr(character), end = "")
        # check if the character is lowercase
        elif character.islower():
            character = ((ord(character) - ord('a') + (key[i])) % 26) + ord('a')
            print(chr(character), end = "")
    else:
        print(character, end="")

    i = (i + 1) % keyLength

print("")